### PR TITLE
add cloud_replicate.Jenkinsfile

### PR DESCRIPTION
**Pull request by dustymabe**

### CODE REVIEW

File: jobs/release.Jenkinsfile [hunk 1]
```diff
@@ -117,6 +117,10 @@ lock(resource: "release-${params.STREAM}", extra: locks) {
             }
         }
 
+        // Update description based on updated set of architectures
+        build_description += "[${params.STREAM}][${basearches.join(' ')}][${params.VERSION}]"
```
Comment (jlebon) [inline jobs/release.Jenkinsfile:121]: Shouldn't this be `=`?

File: jobs/cloud_replicate.Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                echo "ERROR: Some requested architectures did not successfully build"
+                echo "ERROR: Detected built architectures: $builtarches"
+                echo "ERROR: Requested base architectures: $basearches"
+                currentBuild.result = 'FAILURE'
+                return
+            }
+        }
+
+        // Update description based on updated set of architectures
+        build_description += "[${params.STREAM}][${basearches.join(' ')}][${params.VERSION}]"
```
Comment (jlebon) [inline jobs/cloud_replicate.Jenkinsfile:112]: Here too.

File: jobs/cloud_replicate.Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+def stream_info = pipecfg.streams[params.STREAM]
+
+build_description += "[${basearches.join(' ')}][${params.VERSION}]"
+currentBuild.description = "${build_description} Waiting"
+
+// We just lock here out of an abundance of caution in case somehow
+// two jobs run for the same stream, but that really shouldn't happen.
+// Also lock version-arch-specific locks to make sure these builds are finished.
+lock(resource: "cloud-replicate-${params.STREAM}") {
```
Comment (jlebon) [inline jobs/cloud_replicate.Jenkinsfile:72]: I don't think we need this lock.

File: jobs/cloud_replicate.Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                --arch=all --url=s3://${s3_stream_dir}/builds \
+                --aws-config-file \${AWS_BUILD_UPLOAD_CONFIG}
+            """)
+        }
+
+        def builtarches = shwrapCapture("""
+                          cosa shell -- cat builds/builds.json | \
+                              jq -r '.builds | map(select(.id == \"${params.VERSION}\"))[].arches[]'
+                          """).split() as Set
+        assert builtarches.contains("x86_64"): "The x86_64 architecture was not in builtarches."
```
Comment (jlebon) [inline jobs/cloud_replicate.Jenkinsfile:97]: Minor: feels like we could split this out into a shared function.

File: jobs/cloud_replicate.Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                def arch_args = basearches.collect{"--arch ${it}"}.join(" ")
+                shwrap("""
+                cosa generate-release-meta --build-id ${params.VERSION} --workdir .
+                cosa buildupload --build=${params.VERSION} --skip-builds-json \
+                    ${arch_args} s3 --aws-config-file=\${AWS_BUILD_UPLOAD_CONFIG} \
+                    --acl=public-read ${s3_stream_dir}/builds
+                """)
+
+            }
+
```
Comment (jlebon) [inline jobs/cloud_replicate.Jenkinsfile:133]: Extraneous empty lines.

File: jobs/cloud_replicate.Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        currentBuild.result = 'SUCCESS'
+        currentBuild.description = "${build_description} ✓"
+
+// main try finishes here
+} catch (e) {
+    currentBuild.result = 'FAILURE'
+    throw e
+} finally {
+    if (currentBuild.result != 'SUCCESS') {
+        pipeutils.trySlackSend(color: 'danger', message: ":fcos: :bullettrain_front: :trashfire: cloud-replicate <${env.BUILD_URL}|#${env.BUILD_NUMBER}> [${params.STREAM}][${basearches.join(' ')}] (${params.VERSION})")
```
Comment (jlebon) [inline jobs/cloud_replicate.Jenkinsfile:145]: Let's use a different emoji from the release job one? Maybe e.g. `:cloud:`?

File: jobs/release.Jenkinsfile (hunk 1 above)
Comment (dustymabe) [reply jobs/release.Jenkinsfile:121]: will fix

File: jobs/cloud_replicate.Jenkinsfile (hunk 2 above)
Comment (dustymabe) [reply jobs/cloud_replicate.Jenkinsfile:112]: will fix

File: jobs/cloud_replicate.Jenkinsfile (hunk 3 above)
Comment (dustymabe) [reply jobs/cloud_replicate.Jenkinsfile:72]: you're probably right, but we definitely wouldn't want two jobs with same stream/version to run at the same time (on accident), right?

File: jobs/cloud_replicate.Jenkinsfile (hunk 4 above)
Comment (dustymabe) [reply jobs/cloud_replicate.Jenkinsfile:97]: yeah maybe later.

File: jobs/cloud_replicate.Jenkinsfile (hunk 6 above)
Comment (dustymabe) [reply jobs/cloud_replicate.Jenkinsfile]: updated to match what jenkins prints for the stage name for the cloud replication.

File: jobs/cloud_replicate.Jenkinsfile (hunk 5 above)
Comment (dustymabe) [reply jobs/cloud_replicate.Jenkinsfile]: fixed

**Comment by dustymabe** [comment]
Added a new commit that updates emoji use in for slack notifications.

File: jobs/cloud_replicate.Jenkinsfile (hunk 3 above)
Comment (jlebon) [reply jobs/cloud_replicate.Jenkinsfile:72]: Same version, no. Same stream... I guess it's uncommon but perfectly safe, so I wouldn't go out of my way to block it. E.g. maybe you've already started a replication job for a version but then another build comes through that supersedes it but the replicate job isn't done yet and it's too late to cancel it without leaking resources.

I guess the simplest patch is to do `s/STREAM/VERSION/`.

Comment (dustymabe) [reply jobs/cloud_replicate.Jenkinsfile]: done

**Review by jlebon — APPROVED** [review]

**Pull request merged**